manager/controlapi/secret_test.go

```diff
@@ -0,0 +1,322 @@
+package controlapi
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"testing"
+
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/manager/state/store"
+	"github.com/stretchr/testify/assert"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+)
+
+func createSecretSpec(name string, data []byte, labels map[string]string) *api.SecretSpec {
+	return &api.SecretSpec{
+		Annotations: api.Annotations{Name: name, Labels: labels},
+		Data:        data,
+	}
+}
+
+func validateChecksum(t *testing.T, secret *api.Secret, data []byte) {
+	checksumBytes := sha256.Sum256(data)
+	assert.Equal(t, "sha256:"+hex.EncodeToString(checksumBytes[:]), secret.Digest)
+	assert.EqualValues(t, len(data), secret.SecretSize)
+}
+
+func TestValidateSecretSpec(t *testing.T) {
+	type BadServiceSpec struct {
+		spec *api.ServiceSpec
+		c    codes.Code
+	}
+
+	for _, badName := range []string{
+		"",
+		".",
+		"-",
+		"_",
+		".name",
+		"name.",
+		"-name",
+		"name-",
+		"_name",
+		"name_",
+		"/a",
+		"a/",
+		"a/b",
+		"..",
+		"../a",
+		"a/..",
+		"withexclamation!",
+		"with space",
+		"with\nnewline",
+		"with@splat",
+		"with:colon",
+		"with;semicolon",
+		"snowman☃",
+		"o_______________________________________________________________o", // exactly 65 characters
```

Comment:
Aww, I liked the giant frowny face.  :)  But yes `strings.Repeat` would make it clearer, thanks.